tasty-plutus/src/Test/Tasty/Plutus/Internal/TestScript.hs

```diff
@@ -0,0 +1,288 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+module Test.Tasty.Plutus.Internal.TestScript (
+  -- * Types
+  TestScript (..),
+  WrappedValidator (..),
+  WrappedMintingPolicy (..),
+
+  -- * Utility functions
+  mkTestValidator,
+  mkTestValidatorUnsafe,
+  mkTestMintingPolicy,
+  mkTestMintingPolicyUnsafe,
+  toTestMintingPolicy,
+  toTestValidator,
+) where
+
+import Data.Kind (Type)
+import Ledger.Typed.Scripts (Validator)
+import Plutus.V1.Ledger.Api (
+  MintingPolicy,
+  mkMintingPolicyScript,
+  mkValidatorScript,
+ )
+import PlutusTx (CompiledCode, applyCode)
+import PlutusTx.Builtins (BuiltinData, BuiltinString, appendString, trace)
+import PlutusTx.IsData.Class (FromData (fromBuiltinData))
+import PlutusTx.TH (compile)
+import Test.Tasty.Plutus.Internal.Context (Purpose (ForMinting, ForSpending))
+
+{- | Typed wrapper for the 'Validator' and 'MintingPolicy' used to match
+ the datum and redeemer types of the 'Validator' and the data passed to it.
+
+ We don't expose constructors. To create you need to use helper functions
+ such as 'mkTestValidator' and 'mkTestMintingPolicy'. In case you intend
+ to test something tricky, you can use 'mkTestValidatorUnsafe'
+ and 'mkTestMintingPolicyUnsafe' to create a 'TestScript'
+ that accepts datum and/or redeemer inconsistent with its internal type.
+
+ @since 6.0
+-}
+data TestScript (p :: Purpose) where
+  -- | since 6.0
+  TestValidator ::
+    forall (d :: Type) (r :: Type).
+    {getTestValidator :: Validator} ->
+    TestScript ( 'ForSpending d r)
+  -- | since 6.0
+  TestMintingPolicy ::
+    forall (r :: Type).
+    {getTestMintingPolicy :: MintingPolicy} ->
+    TestScript ( 'ForMinting r)
+
+-- | @since 6.0
+deriving stock instance Show (TestScript p)
+
+{- | A wrapper for an untyped 'Validator'. This is similar
+ to 'WrappedValidatorType' from the Plutus 'Ledger.Typed.Scripts' module.
+ It`s aim is to force you to use 'toTestValidator' to create a 'TestValidator'.
+
+ @since 6.0
+-}
+newtype WrappedValidator
+  = WrappedValidator (BuiltinData -> BuiltinData -> BuiltinData -> ())
+
+{-# INLINEABLE getWrappedValidator #-}
+getWrappedValidator ::
+  WrappedValidator -> (BuiltinData -> BuiltinData -> BuiltinData -> ())
+getWrappedValidator (WrappedValidator v) = v
+
+{- | A wrapper for an untyped 'MintingPolicy'. This is similar
+ to 'WrappedMintingPolicyType' from the Plutus 'Ledger.Typed.Scripts' module.
+ It`s aim is to force you to use 'toTestMintingPolicy' to create
+ a 'TestMintingPolicy'.
+
+ @since 6.0
+-}
+newtype WrappedMintingPolicy
+  = WrappedMintingPolicy (BuiltinData -> BuiltinData -> ())
+
+{-# INLINEABLE getWrappedMintingPolicy #-}
+getWrappedMintingPolicy ::
+  WrappedMintingPolicy -> (BuiltinData -> BuiltinData -> ())
+getWrappedMintingPolicy (WrappedMintingPolicy mp) = mp
+
+{- | 'mkTestValidator' provide the way for creating 'TestScript' with
+ the given 'Validator', parameterized with proper datum and redeemer types.
```

Comment:
Better phrasing:

Create a 'TestScript' with the given 'Validator', parameterized with proper datum and redeemer types.